File: js/src/hash.js
// @flow

// Copyright 2016 The Noms Authors. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

import sha1 from './sha1.js';

export const sha1Size = 20;
const pattern = /^sha1-[0-9a-f]{40}$/;

const sha1Prefix = 'sha1-';
const sha1PrefixLength = sha1Prefix.length;

function uint8ArrayToSha1(a: Uint8Array): string {
  const sha1 = new Array(1 + sha1Size * 2);
  sha1[0] = [sha1Prefix];
  for (let i = 0; i < a.length; i++) {
    sha1[i + 1] = byteToAscii[a[i]];
  }
  return sha1.join('');
}

function sha1ToUint8Array(s: string): Uint8Array {
  const digest = new Uint8Array(sha1Size);
  for (let i = 0; i < sha1Size; i++) {
    const hc = asciiToBinary(s.charCodeAt(sha1PrefixLength + 2 * i));
    const lc = asciiToBinary(s.charCodeAt(sha1PrefixLength + 2 * i + 1));
    digest[i] = hc << 4 | lc;
  }
  return digest;
}

export default class Hash {
  _digest: Uint8Array;

  constructor(digest: Uint8Array) {
    // Make a copy to prevent holding on to the data that was passed in.
    this._digest = new Uint8Array(digest);
  }

  get digest(): Uint8Array {
    return this._digest;
  }

  isEmpty(): boolean {
    for (let i = 0; i < sha1Size; i++) {
      if (this._digest[i]) {
        return false;
      }
    }
    return true;
  }

  equals(other: Hash): boolean {
    for (let i = 0; i < sha1Size; i++) {
      if (this._digest[i] !== other._digest[i]) {
        return false;
      }
    }
    return true;
  }

  compare(other: Hash): number {
    for (let i = 0; i < sha1Size; i++) {
      const d = this._digest[i] - other._digest[i];
      if (d) {
        return d;
      }
    }
    return 0;
  }

  toString(): string {
    return uint8ArrayToSha1(this._digest);
  }

  static parse(s: string): Hash {
    if (!pattern.test(s)) {
      throw new Error(`Could not parse hash: ${s}`);
    }
    return new Hash(sha1ToUint8Array(s));
  }

  static maybeParse(s: string): ?Hash {
    if (pattern.test(s)) {
      return new Hash(sha1ToUint8Array(s));
    }
    return null;
  }

  static fromData(data: Uint8Array): Hash {
    return new Hash(sha1(data));
  }
}

export const emptyHash = new Hash(new Uint8Array(sha1Size));

function asciiToBinary(cc: number): number {
  // This only accepts the char code for '0' - '9', 'a' - 'f'
  return cc - (cc <= 57 ? 48 : 87); // '9', '0', 'a' - 10
}

// Precompute '00' to 'ff'.
const byteToAscii = new Array(256);
for (let i = 0; i < 256; i++) {
  byteToAscii[i] = (i < 0x10 ? '0' : '') + i.toString(16);
}

File: go/hash/hash.go
// Copyright 2016 The Noms Authors. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

package hash

import (
	"bytes"
	"crypto/sha512"
	"fmt"

	"github.com/attic-labs/noms/go/d"
)

const (
	HashSize = 32
)

var (
	emptyHash = Hash{}
)

// The hash of a Noms value.
//
// Noms serialization version 1 uses the first 32 bytes of sha512. We could use sha512/256, the
// official standard for a half-size sha512, but it is much less commonly used, meaning that there
// is less likelihood of having a good fast implementation on any given platform. For example, at
// time of writing, there is no native implementation in Node.
//
// sha512 was chosen because:
// - sha1 is no longer recommended
// - blake is not commonly used, not a lot of platform support
// - sha3 is brand new, no library support
// - within sha2, sha512 is faster than sha256 on 64 bit
// - we don't need 512 bit hashes - we don't really even need 256 bit, but 256 is very common and
//   leads to a nice round number of base36 digits: 50.
//
// In Noms, the hash function is a component of the serialization version, which is constant over
// the entire lifetime of a single database. So clients do not need to worry about encountering
// multiple hash functions in the same database.
type Hash struct {
	digest Digest
}

type Digest [HashSize]byte

// Digest returns a *copy* of the digest that backs Hash.
func (r Hash) Digest() Digest {
	return r.digest
}

func (r Hash) IsEmpty() bool {
	return r.digest == emptyHash.digest
}

// DigestSlice returns a slice of the digest that backs A NEW COPY of Hash, because the receiver of this method is not a pointer.
func (r Hash) DigestSlice() []byte {
	return r.digest[:]
}

func New(digest Digest) Hash {
	return Hash{digest}
}

func FromData(data []byte) Hash {
	r := sha512.Sum512(data)
	d := Digest{}
	copy(d[:], r[:HashSize])
	return New(d)
}

// FromSlice creates a new Hash backed by data, ensuring that data is an acceptable length.
func FromSlice(data []byte) Hash {
	d.Chk.True(len(data) == HashSize)
	digest := Digest{}
	copy(digest[:], data)
	return New(digest)
}

func Parse(s string) Hash {
	r, ok := MaybeParse(s)
	if !ok {
		d.Exp.Fail(fmt.Sprintf("Cound not parse Hash: %s", s))
	}
	return r
}

func (r Hash) Less(other Hash) bool {
	return bytes.Compare(r.digest[:], other.digest[:]) < 0
}

func (r Hash) Greater(other Hash) bool {
	return bytes.Compare(r.digest[:], other.digest[:]) > 0
}
